Change how to write 2-dimensional table-tests (#637)
kyma-project/community: concepts/eventing-code-guidelines.md

</details>

<details>
  <summary>2-dimensional table-driven test</summary>

> There may be situation when you need a table-driven test with more than one dimension. The following shows a best practice example for a 2-dimensional table-driven test.

Reviewer: ```suggestion
There may be situations when you need a table-driven test with more than one dimension. See the following best practice example for a 2-dimensional table-driven test:
```

```go
func TestTwoDimensions(t *testing.T) {
	// first dimension
	testCases := []struct {
		name               string
		givenSender        CESender
		wantHTTPStatusCode int
	}{
		{
			name:        "binary cloud event sender",
			givenSender: myCEBinarySender,
		},
	}
	// second dimension
	cloudEvents := []struct {
		name, givenCEType string
	}{
		{
			name:               "proper cloud event",
			givenCEType:        "order.created.v1",
			wantHTTPStatusCode: http.StatusOK,
		},
	}
	for _, tc := range testCases {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {

> 			for _, ce := range ce {

Reviewer 2: ```suggestion
			for _, ce := range cloudEvents {
```

				ce := ce
				t.Run(ce.name, func(t *testing.T) {
					t.Parallel()
					res := functionUnderTest(tc.givenSender, ce.givenCEType)
					// check res == ce.wantHTTPStatusCode
				})
			}
		})
	}
}
```

<details>
  <summary>Reason for variable reassignment (tc := tc)</summary>

The loop iteration variable in Go is a single variable. The closure for the second t.Run is executed inside a goroutine. 
If there are multiple entries in the `cloudEvents` struct, `ce` references the last entry in `cloudEvents` in every iteration. The problem occurs because the closure referring to `tc` and `ce` is not executed in **sync** with the **for loop** (because of t.parallel).
To prevent this problem, `ce` and `tc` must be copied (`ce := ce`).

> The linter [scopelint](https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/scopelint.go) warns about the possible problem whenever `ce` or `testCase` is used.

Reviewer 2: ```suggestion
The linter [scopelint](https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/scopelint.go) warns about the possible problem whenever `ce` or `tc` is used.
```

> Adding `// nolint:scopelint` to silence scopelint must be used with caution.

Reviewer: What bad things can happen if you use `// nolint:scopelint` and choose to get no warning? I'd suggest a warning like this:

```suggestion
>**CAUTION:** If you add `// nolint:scopelint` to silence scopelint, you might not notice when `ce` references {a wrong/different??} entry. Use it with caution and make sure to check {...?}.
```

When not using t.Run in a nested way, the test could look like this:

```go
// ...
for _, tc := range testCases {
	tc := tc
	for _, ce := range ce {
		ce := ce
		t.Run(tc.name + " - " + ce.name, func(t *testing.T) { // only this line changed
			t.Parallel()
			res := functionUnderTest(tc.givenSender, ce.givenCEType)
			// check res == ce.wantHTTPStatusCode
		})
	}
}

$ go test -v
=== RUN   TestTwoDimensions
=== RUN   TestTwoDimensions/binary_cloud_event_sender_-_proper_cloud_event
--- PASS: TestTwoDimensions (0.00s)
    --- PASS: TestTwoDimensions/binary_cloud_event_sender_-_proper_cloud_event (0.00s)
PASS
ok      test    0.199s
```

The output of the test with nested t.Run looks like this:

```shell
$ go test -v
=== RUN   TestTwoDimensions
=== RUN   TestTwoDimensions/binary_cloud_event_sender
=== RUN   TestTwoDimensions/binary_cloud_event_sender/proper_cloud_event
--- PASS: TestTwoDimensions (0.00s)
    --- PASS: TestTwoDimensions/binary_cloud_event_sender (0.00s)
        --- PASS: TestTwoDimensions/binary_cloud_event_sender/proper_cloud_event
 (0.00s)
PASS
ok      test    0.182s
```

> When you look at the output that both examples produce, you can see that the test name is different (`binary_cloud_event_sender_-_proper_cloud_event` vs `binary_cloud_event_sender/proper_cloud_event`). Each subtest will add `/<test_name>` to the test name.

Reviewer: ```suggestion
When you look at the output of both examples, you can see that the test name is different: (`binary_cloud_event_sender_-_proper_cloud_event` vs `binary_cloud_event_sender/proper_cloud_event`). Each subtest adds `/<test_name>` to the test name.
```

> The **advantages** of using t.Run in a nested way are that:

Reviewer 2: ```suggestion
The **advantages** of using t.Run in a nested way are:
```

> The **advantage** of using t.Run in a nested way is that:

Reviewer: ```suggestion
The **advantages** of using t.Run in a nested way are that:
```

		},
	}
	for _, tc := range testCases {
	  t.Run(tc.name, func(t *testing.T) {
		for _, ceTestCase := range handlertest.TestCasesForCloudEvents {
			t.Run(ceTestCase.Name, func(t *testing.T) {
			// other code is unchanged
			})
		}
	})
}
```

<details>
  <summary>Reason for using nested t.Run</summary>

Let us assume that `tc.name` is equal to `parent test name` and `ceTestCase.Name` is equal to `child test name`.

When you look at the output that both examples produce, you can see that the test name is different (`parent_test_name_-_child_test_name` vs `parent_test_name/child_test_name`). Each subtest will add `/<test_name>` to the test name.
The **advantage** of using t.Run in a nested way is that:

> 1. The test name is easier to read (`parent_test_name/child_test_name`).

Reviewer: if the order is not important, use a bullet list
```suggestion
- The test name is easier to read (`parent_test_name/child_test_name`).
```
